### PR TITLE
[MRG] Cleaning of successive halving

### PR DESCRIPTION
**Pull request by NicolasHug**
Various little things:

- Use stratified subsampling
- `scoring` param cannot accept lists or dicts. This is because when doing multi-metric evaluation, BaseSearchCV requires `refit` to be a string, and we cannot allow this since we need to override it.
- refit only accepts True or False
- raise error if max_budget isn't set when budget_on!=n_samples
- removed f strings

### CODE REVIEW

**Comment by amueller** [comment]
thanks!

File: dabl/search.py [hunk 1]
```diff
@@ -199,19 +205,14 @@ def _run_search(self, evaluate_candidates, X, y, groups):
...
             if self.budget_on == 'n_samples':
-                # XXX FIXME TODO
-                # subsampling should be stratified. We can't use
-                # train_test_split because it complains about testset being too
-                # small in some cases
-                indexes = rng.choice(X.shape[0], r_i, replace=False)
-                X_iter = safe_indexing(X, indexes)
-                y_iter = safe_indexing(y, indexes)
+                stratify = y if is_classifier(self.estimator) else None
+                X_iter, y_iter = resample(X, y, replace=False,
```
Comment (amueller) [inline dabl/search.py:214]: aren't you missing an ``n_samples`` parameter here?

Comment (NicolasHug) [reply dabl/search.py:214]: yyuuup :/

should add `n_samples=r_i`...

Sorry about that

Comment (amueller) [reply dabl/search.py:214]: I'll push the fix but could you think about a regression test for this please?

Comment (NicolasHug) [reply dabl/search.py:214]: We have checks that make sure the `r_i` are corrects. Making sure `X_iter.shape[0]` is correct is more difficult and can only be done indirectly, e.g. checking stuff like `results.groupby(['iter']).mean_fit_time.mean()` and asserting that this grows... But it's subject to side effects.

Or we could add a private list attribute `_n_samples_iter`?
